csharp/src/Ice/Communicator.cs

```diff
@@ -1123,13 +1113,11 @@ public void Destroy()
             return adminAdapter.CreateProxy(adminIdentity, IObjectPrx.Factory);
         }
 
-        /// <summary>
-        /// Check whether the communicator has been shut down.
-        /// </summary>
+        /// <summary>Check whether the communicator has been shut down.</summary>
         /// <returns>True if the communicator has been shut down; false otherwise.</returns>
         public bool IsShutdown()
         {
-            lock (this)
+            lock (_mutex) // TODO do we need this lock, isn't _isShutdown assignment atomic?
```

Comment:
Reading and writing booleans is indeed atomic, so I agree we don't need this lock.

But there still could be an issue of thread visibility, changes in one thread might not appear on others due to caching. So we should probably mark `_isShutdown` as
```
private volatile bool _isShutdown;
```
to disable thread-local caching. Or leave the lock, either or.

Reply:
Keeped the TODO comment for now